lib/sda/overlay.go

```diff
@@ -298,6 +336,55 @@ func (o *Overlay) Close() {
 		dbg.Lvl4("Closing node", n.TokenID())
 		o.nodeDelete(n.Token())
 	}
+
+	o.instancesLock.Lock()
+	for _, tni := range o.instances {
+		tni.Close()
+	}
+}
+
+func (o *Overlay) newTreeNodeInstance(t *Tree, tn *TreeNode, servID ServiceID) *TreeNodeInstance {
+	tok := &Token{
+		TreeNodeID:   tn.Id,
+		TreeID:       t.Id,
+		EntityListID: t.EntityList.Id,
+		ServiceID:    servID,
+		RoundID:      RoundID(uuid.NewV4()),
+	}
+
+	tni := newTreeNodeInstance(o, tok, tn)
+	o.instancesLock.Lock()
+	defer o.instancesLock.Unlock()
+	o.instances[tok.Id()] = tni
+
+	o.RegisterTree(t)
+	o.RegisterEntityList(t.EntityList)
+	dbg.Lvl4(o.host.workingAddress, "Registered NewTreeNodeInstance with Tree + EntityList + ServiceID")
+	return tni
+}
+
+var ErrWrongTreeNodeInstance = errors.New("TreeNodeInstance associated with this ProtocolInstance is already registered")
+var ErrProtocolRegistered = errors.New("A ProtocolInstance already has been registered using this TreeNodeInstance!")
```

Comment:
It's for the Service so they know what is wrong instead of just seeing an error text. They can do:
```
if err := c.RegisterProtocolInstance(pi); err == sda.ErrProtocolRegistered {
  // act accordingly
} else err == sda.ErrWrongTreeNodeInstance {
  // act accordingly 
} else {
  // drop everything and run away
}
```
Just for convenience, but not really important.